src/utils/src/net/tap.rs

```diff
@@ -543,40 +543,11 @@ mod tests {
 
     // Given a buffer of appropriate size, this fills in the relevant fields based on the
     // provided information. Payload refers to the UDP payload.
-    fn build_packet(
-        buf: &mut [u8],
-        src_mac: MacAddr,
-        dst_mac: MacAddr,
-        payload: &[u8],
-        src_addr: &SocketAddrV4,
-        dst_addr: &SocketAddrV4,
-    ) {
-        let mut eth = EthernetFrame::from_bytes(buf).unwrap();
-
-        eth.set_src_mac(src_mac)
-            .set_dst_mac(dst_mac)
-            .set_ethertype(ETHERTYPE_IPV4);
-
-        let ip_header_len_bytes = 20; // 20 bytes for header (min length)
-        let mut ipv4 = IPv4Packet::write_header(
-            eth.payload_mut(),
-            PROTOCOL_UDP,
-            *src_addr.ip(),
-            *dst_addr.ip(),
-        )
-        .unwrap()
-        .with_header_and_payload_len_unchecked(
-            ip_header_len_bytes,             // IHL = 5
-            UDP_HEADER_SIZE + payload.len(), // Udp packet
-            true,
-        );
-
-        let udp = UdpDatagram::write_incomplete_datagram(ipv4.payload_mut(), payload).unwrap();
-        udp.finalize(
-            src_addr.port(),
-            dst_addr.port(),
-            Some((*src_addr.ip(), *dst_addr.ip())),
-        );
+    fn build_packet(payload: &[u8]) -> Vec<u8> {
+        let mut base = Vec::new();
```

Comment:
Can you add some comments to what these bytes represent?

For example, you can use `#[rufmt::skip]` on the initialization of base and then split the content per logical component. 
```rust
base = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 18, 52, 86, 120, 154, 189, // source port
            ...............         // destination port
            ...................... // payload length
            .......
]
```

BTW, shouldn't the payload length be written based on the `payload` parameter size?

Reply:
I separated this ugly array into multiple variables with each representing a particular segment of the Ethernet, IP, UDP headers. I also added some comments explaining each variable. 
https://github.com/firecracker-microvm/firecracker/blob/72d58ff43065e8188b51e65600f3fb9e4bc56917/src/utils/src/net/tap.rs#L447-L494